Application passwords implementation

scabrero opened the pull request:
This pull request replaces #2457 and follows the design document #2427.

It is just the server side for now.

Related to #41 

Checklist

- [X] This pr contains no AI generated code
- [X] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

scabrero commented:
@Firstyear, the last test for:

```
Since application passwords are related to applications, on delete of an application all entries
that have a bound application password should be removed from user accounts.
```

fails. Any hint about how to enforce reference integrity on a "complex" attribute?

Firstyear commented on book/src/access_control/intro.md:
```diff
@@ -61,6 +61,7 @@ tasks within Kanidm as required.
 | `idm_schema_admins`          | add and modify elements of schema                                       |
 | `idm_service_account_admins` | create and modify service accounts                                      |
 | `idm_unix_admins`            | enable posix attributes on accounts and groups                          |
+| `idm_application_admins`     | create and modify applications
```
@yaleman Thoughts?

Firstyear commented on server/lib/src/idm/ldap.rs:
```diff
@@ -525,6 +508,63 @@ impl LdapServer {
+                    let app = app.unwrap().as_str();
+
+                    if app.is_empty() {
+                        // That's weird ...
```
As above, log what and why

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/idm/ldap.rs:
```diff
@@ -525,6 +508,63 @@ impl LdapServer {
+            }
+        }
+
+        Err(OperationError::NoMatchingEntries)
```
Error message explaining how this occured and the case.

scabrero replied:
Done.

Firstyear commented on server/lib/src/idm/mod.rs:
```diff
@@ -5,6 +5,7 @@
 
 pub mod account;
 pub(crate) mod accountpolicy;
+pub mod application;
```
pub(crate), doesn't need to be public to everything.

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/idm/server.rs:
```diff
@@ -2095,6 +2119,11 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
     pub fn commit(mut self) -> Result<(), OperationError> {
+        if self.qs_write.get_changed_app() {
+            self.qs_write
+                .get_apprs_set()
```
get_apprs_set probably needs a more descriptive name. I think you copy-pasted oauth2rs here, and may not have realised that the "rs" stands for "resource server". So here you probably don't want "apprs", you want something descriptive like "get_ldap_application_set".

scabrero replied:
All names corrected and without abbreviations.

Firstyear commented on server/lib/src/idm/server.rs:
```diff
@@ -2095,6 +2119,11 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
+        if self.qs_write.get_changed_app() {
+            self.qs_write
+                .get_apprs_set()
+                .and_then(|apprs_set| self.apprs.reload(apprs_set))?;
```
As above, we are trying to move away from abreviated names if possible.

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/idm/server.rs:
```diff
@@ -2150,12 +2179,53 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
+                &ev.ident,
+            )
+            .map_err(|e| {
+                request_error!(error = ?e);
```
error!

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/server/migrations.rs:
```diff
@@ -933,6 +933,7 @@ impl<'a> QueryServerWriteTransaction<'a> {
             SCHEMA_ATTR_ACCOUNT_EXPIRE.clone().into(),
             SCHEMA_ATTR_ACCOUNT_VALID_FROM.clone().into(),
             SCHEMA_ATTR_API_TOKEN_SESSION.clone().into(),
+            SCHEMA_ATTR_APPLICATION_PASSWORD.clone().into(),
```
see migrations comment above, this doesn't go here. It needs to be in the domain level 5 to 6 migration https://github.com/kanidm/kanidm/pull/2576

Firstyear commented on server/lib/src/server/migrations.rs:
```diff
@@ -1124,6 +1127,9 @@ impl<'a> QueryServerWriteTransaction<'a> {
             IDM_ACP_SERVICE_ACCOUNT_ENTRY_MANAGED_BY_MODIFY_V1.clone(),
             IDM_ACP_HP_SERVICE_ACCOUNT_ENTRY_MANAGED_BY_MODIFY_V1.clone(),
             IDM_ACP_SERVICE_ACCOUNT_MANAGE_V1.clone(),
+            IDM_ACP_APPLICATION_CREATE_V1.clone(),
```
As above.

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/server/mod.rs:
```diff
@@ -901,13 +904,27 @@ pub trait QueryServerTransaction<'a> {
             })
     }
 
+    fn get_apprs_set(&mut self) -> Result<Vec<Arc<EntrySealedCommitted>>, OperationError> {
```
Why do you have this twice? with get_apprs and get_application_rs_set?

scabrero replied:
A mistake, fixed.

Firstyear commented on server/lib/src/valueset/apppwd.rs:
```diff
@@ -0,0 +1,210 @@
+
+#[derive(Debug, Clone)]
+pub struct ValueSetApplicationPassword {
+    map: BTreeMap<Uuid, ApplicationPassword>,
```
comment what the uuid here is.

scabrero replied:
Done.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+            }
+        };
+
+        // TODO: Should we check if the group still exists or just trust in ref. integrity?
```
ref integrity is trustworthy.

scabrero replied:
Ok, comment removed.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+
+        if !within_valid_window {
+            security_info!("Account has expired or is not yet valid, not allowing to proceed");
+            return Err(OperationError::SessionExpired);
```
Do I use session expired else where? Or account expired? 🤔

scabrero replied:
`SessionExpired`

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+
+        // TODO: Should we check if the group still exists or just trust in ref. integrity?
+
+        if !account.groups.iter().any(|e| e.uuid() == linked_group_uuid) {
```
Probably easier to just check the entries memberOf directly? I think there is a better way than an iter() here.

scabrero replied:
Ok, fixed.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+    const TEST_CURRENT_TIME: u64 = 6000;
+
+    #[idm_test]
+    async fn test_idm_application_excludes(idms: &IdmServer, _idms_delayed: &mut IdmServerDelayed) {
```
In general, more comments to the logic of what and why you are testing things is needed. See https://github.com/kanidm/kanidm/blob/master/server/lib/src/repl/tests.rs#L196 for example.

scabrero replied:
Ok, comments added.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+    }
+
+    #[idm_test]
+    async fn test_idm_application_no_linked_group(
```
I thought linked group was required?

Firstyear replied:
Nope it wasn't commented on schema that it needs to be a must attribute.

scabrero replied:
Discussion continues below.

Firstyear commented on server/lib/src/constants/schema.rs:
```diff
@@ -1055,4 +1078,16 @@ pub static ref SCHEMA_CLASS_OAUTH2_RS_PUBLIC_DL4: SchemaClass = SchemaClass {
+
+    description: "The class representing an application".to_string(),
+    systemmust: vec![Attribute::Name.into()],
+    systemmay: vec![Attribute::Description.into(), Attribute::LinkedGroup.into()],
```
systemmust for linked group - an application with no linked group doesnt't make sense.

scabrero replied:
Reading the design doc I interpret two steps, first the application is created, then admin establishes the `app <-> group` association. But we can can change the design document :)

Firstyear replied:
I think we need to create the group association as we create the application.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,631 @@
+            assert!(app.linked_group.is_some_and(|u| u == test_grp_uuid));
+        }
+
+        // Test reference integrity. If group is removed the attribute must go
```
If you attempt to delete the group, the linked group presence *blocks* the group being deleted, showing the user that something still requires it! You have to delete the application *first* then you can delete the group.

scabrero replied:
Ok, test updated.

Firstyear commented on server/lib/src/idm/ldap.rs:
```diff
@@ -823,6 +865,591 @@ mod tests {
+
+            assert!(idms_prox_write.commit().is_ok());
+
+            // event silently dropped on duplicated label (check Entry::add_ava())
```
Rather than silent drop, I think it should either error or overwrite. Silently losing a change is user hostile.

scabrero replied:
I changed the code to overwrite on duplicated label for now.

Firstyear commented on server/lib/src/valueset/apppwd.rs:
```diff
@@ -0,0 +1,210 @@
+    map: BTreeMap<Uuid, ApplicationPassword>,
+}
+
+impl ValueSetApplicationPassword {
```
I think you need to refer to the design doc 

https://github.com/kanidm/kanidm/blob/master/book/src/developers/designs/application_passwords.md#accounts

Because of the structure you have here, you end up with code like https://github.com/kanidm/kanidm/pull/2578/files#diff-c01cb17e7b8ec9eef1cf7458a914cb61c72427e4377697d916781127a71c2942R714 which has to iterate over every possible password in the account, but in the design process, we specifically addressed this and said "lets make this structure have applicationUuid on the outside, then a multi-value structure nested inside". That way we optimise for the common path, and we get a nice, easy, common group and way to delete refered application passwords. It's a pain, but it means you have to change some of your logic and structure here, but it'll end up forcing the rest of the implementation to be nicer and map to the problem space better.

scabrero replied:
Done, application's uuid is the map index now.

scabrero commented on server/lib/src/be/dbvalue.rs:
```diff
@@ -337,16 +337,6 @@ pub struct DbValueCredV1 {
     pub data: DbCred,
 }
 
-#[derive(Serialize, Deserialize, Debug)]
```
Do you mean `enum DbApiToken`?  Because it looks like a leftover, it is not used anywhere.

scabrero commented on server/lib/src/constants/acp.rs:
```diff
@@ -1693,16 +1696,19 @@ lazy_static! {
             Attribute::PrimaryCredential,
             Attribute::PassKeys,
             Attribute::AttestedPasskeys,
+            Attribute::ApplicationPassword,
```
IMO it depends on how you interpret `idm_acp_people_credential_reset` control.

In a strict way an application password is another kind of credential, so anyone entitled to reset credentials should be able to delete them, but there is no point on allowing modifications until we have the "enabled" boolean.

On the other hand if you think about `idm_acp_people_credential_reset` as the control used to just reset the credentials when a user forgets the password I agree leaving application passwords out.

Firstyear replied:
I think it's justu for the "forgot password" so I think we leave applications passwords out.

scabrero replied:
Ok, removed.

scabrero commented on server/lib/src/constants/acp.rs:
```diff
@@ -1969,3 +1978,91 @@ lazy_static! {
+        create_classes: vec![EntryClass::Object, EntryClass::Application],
+        ..Default::default()
+    };
+    pub static ref IDM_ACP_APPLICATION_MANAGE_V1: BuiltinAcp = BuiltinAcp {
```
Creating and managing applications does not fit in any of the existing ACPs, except maybe:
* `IDM_ACP_SERVICE_ACCOUT_*` ?
* `IDM_ACP_DOMAIN_ADMIN` ?

Firstyear replied:
I think we need some more thought here about this. For example what are the administrator use cases? Do we need to delegate rights? 

We should look at this as "how will admins and users interact" and use that to create access controls instead.

scabrero replied:
Conversation continues below.

scabrero commented on server/lib/src/constants/schema.rs:
```diff
@@ -638,6 +638,27 @@ pub static ref SCHEMA_ATTR_LIMIT_SEARCH_MAX_FILTER_TEST_DL6: SchemaAttribute = S
+pub static ref SCHEMA_ATTR_LINKED_GROUP: SchemaAttribute = SchemaAttribute {
+    uuid: UUID_SCHEMA_ATTR_LINKED_GROUP,
+    name: Attribute::LinkedGroup.into(),
+    description: "The application linked group".to_string(),
```
the description or the attribute name?

Firstyear replied:
description.

scabrero replied:
Fixed.

scabrero commented on server/lib/src/constants/schema.rs:
```diff
@@ -1055,4 +1078,16 @@ pub static ref SCHEMA_CLASS_OAUTH2_RS_PUBLIC_DL4: SchemaClass = SchemaClass {
+    systemmust: vec![Attribute::Name.into()],
+    systemmay: vec![Attribute::Description.into(), Attribute::LinkedGroup.into()],
+    systemsupplements: vec![EntryClass::ServiceAccount.into()],
+    systemexcludes: vec![EntryClass::Person.into()],
```
Fixed, but Person must exclude application then. Otherwise creating
```
let e1 = entry_init!(
            (Attribute::Class, EntryClass::Object.to_value()),
            (Attribute::Class, EntryClass::Account.to_value()),
            (Attribute::Class, EntryClass::Application.to_value()),
            (Attribute::Class, EntryClass::Person.to_value()),
            (Attribute::Name, Value::new_iname("test_app_name")),
            (Attribute::Uuid, Value::Uuid(test_entry_uuid)),
            (Attribute::Description, Value::new_utf8s("test_app_desc")),
            (
                Attribute::DisplayName,
                Value::new_utf8s("test_app_dispname")
            )
        );
```
succeeds.

scabrero commented on server/lib/src/credential/apppwd.rs:
```diff
@@ -0,0 +1,38 @@
+        label: &str,
+        cleartext: &str,
+    ) -> Result<ApplicationPassword, OperationError> {
+        let p = CryptoPolicy::minimum();
```
Fixed.

Firstyear replied:
Is it just not in the PR yet?

scabrero replied:
Not pushed yet.

scabrero commented on server/lib/src/idm/event.rs:
```diff
@@ -508,3 +528,36 @@ impl AuthResult {
 */
+
+#[derive(Debug)]
+pub struct GenerateApplicationPasswordEvent {
```
Fixed.

scabrero commented on server/lib/src/idm/ldap.rs:
```diff
@@ -525,6 +508,63 @@ impl LdapServer {
+
+                let mut idm_auth = idms.auth().await;
+                let usr_uuid = idm_auth.qs_read.name_to_uuid(usr).map_err(|e| {
+                    request_error!(err = ?e, ?usr, "Error resolving rdn to target");
```
Fixed.

scabrero commented on server/lib/src/idm/ldap.rs:
```diff
@@ -525,6 +508,63 @@ impl LdapServer {
+                    e
+                })?;
+
+                if app.is_some() {
```
Fixed.

scabrero commented on server/lib/src/constants/schema.rs:
```diff
@@ -649,6 +670,7 @@ pub static ref SCHEMA_CLASS_PERSON: SchemaClass = SchemaClass {
     systemmay: vec![
         Attribute::Mail.into(),
         Attribute::LegalName.into(),
+        Attribute::ApplicationPassword.into(),
```
Fixed, new DL6 person.

scabrero commented on server/lib/src/idm/ldap.rs:
```diff
@@ -85,11 +86,18 @@ impl LdapServer {
+            format!("^((?P<attr>[^=,]+)=(?P<val>[^=,]+),)?(app=(?P<app>[^=,]+),)?({basedn})$")
+                .as_str(),
+        )
+        .swap_greed(true)
```
Ok, I have added comments and a new test.

Firstyear commented on server/lib/src/constants/acp.rs:
```diff
@@ -1737,20 +1743,23 @@ lazy_static! {
             Attribute::AccountValidFrom,
             Attribute::PassKeys,
             Attribute::AttestedPasskeys,
+            Attribute::ApplicationPassword,
```
Should these still be here? Per above it's the same logic that this role is for admins to reset other admins who forgot a password rather than needing to change their application passwords.

scabrero replied:
Not pushed yet.

scabrero commented:
Hi @Firstyear, PR is not updated yet because I am rebasing on top of master commit by commit while addressing your comments.

Firstyear commented:
> Hi @Firstyear, PR is not updated yet because I am rebasing on top of master commit by commit while addressing your comments.

Ahhh okay. Sorry about that :)

Want me to wait til you are ready for me to look again then?

scabrero commented:
> > Hi @Firstyear, PR is not updated yet because I am rebasing on top of master commit by commit while addressing your comments.
> 
> Ahhh okay. Sorry about that :)
> 
> Want me to wait til you are ready for me to look again then?

Except the refint for the ApplicationPassword struct to make the last test pass, it is now ready for a second review :)

scabrero commented:
Ready for review.

Firstyear commented on server/lib/src/constants/acp.rs:
```diff
@@ -2064,3 +2067,91 @@ lazy_static! {
+        ],
+        create_classes: vec![EntryClass::Object, EntryClass::Application],
+        ..Default::default()
+    };
```
minor, but try to space things out. When you have structs stacked without spaces it's hard to follow and separate them.

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/constants/acp.rs:
```diff
@@ -2064,3 +2067,91 @@ lazy_static! {
+//   needs to be able to assign to entry managed by
+lazy_static! {
+    pub static ref IDM_ACP_APPLICATION_CREATE_DL6: BuiltinAcp = BuiltinAcp {
+        classes: vec![
```
Previous comment was "how do you envisage these access controls working with human roles and tasks", and I think you need to answer that because these access controls otherwise seem confused.

scabrero replied:
Ok, I can think about two possibilities:

Use case 1, the IDM administrator creates the application, links the group, and sets up the members.
Use case 2, there is a person managing service X, e.g. mail. The IDM administrator creates the app, links a group, 
and delegates group permission for the group to the X service manager who sets up the members.

Firstyear replied:
Okay, so I actually think both possibilities are valid. We have the ability to create new applications (which we will give to an idm_application_admins group, and by default, idm_admins is a member of that, see https://github.com/kanidm/kanidm/blob/master/server/lib/src/constants/groups.rs#L203 and https://github.com/kanidm/kanidm/blob/master/server/lib/src/constants/acp.rs#L594 ) and then for group 2, we already have a delegation feature thanks to our "Entry Managed By" feature. You can see an example of this https://github.com/kanidm/kanidm/blob/master/server/lib/src/constants/acp.rs#L1047 

So we'll need to support  Attribute::EntryManagedBy on the LDAP application class, as well as make EntryManagedBy writable by the application_admins group, then there is a second ACP for ApplicationEntryManager which has the receiver as EntryManager and similar or constrained application rights. Does that seem reasonable?

scabrero replied:
Ok, it is clearer now but I still have questions:

For case 1:
I have merged `IDM_ACP_APPLICATION_{CREATE,MANAGE,DELETE}_DL7` into `IDM_ACP_APPLICATION_MANAGE_DL7`. The receiver is the `idm_application_admins` builtin group and has "full control" over the application entry.

For case 2:
I don't see the need for `Attribute::EntryManagedBy` in the application entry nor the need for a second ACP having `EntryManager` as receiver. IIUC a `idm_application_admins_group` member will create the application and will link a group, and then will delegate the *linked group* management, not the application entry. And as you said we already have `IDM_ACP_GROUP_ENTRY_MANAGER_V1` to allow the delegated person to modify the linked group members. Unless you think the delegated person should be able to modify the linked group in the application entry.

Firstyear replied:
But linked group is about the users who can *access* the application, not the group that *owns and manages* the application integration itself. So that's why I think there needs to be a separation.

scabrero replied:
Ok, understood. I have added it.

Firstyear commented on server/lib/src/constants/schema.rs:
```diff
@@ -638,6 +638,27 @@ pub static ref SCHEMA_ATTR_LIMIT_SEARCH_MAX_FILTER_TEST_DL6: SchemaAttribute = S
+    description: "A reference to the group linked to the entry".to_string(),
+
+    multivalue: false,
+    sync_allowed: true,
```
Do you know what the sync_allowed attribute affects?

scabrero replied:
Attribute or class can be imported from external scim?

Firstyear replied:
So I think the question is if there is a use case here where we would need to import a linked group? I think currently, the answer is no, so we should not set sync_allowed unless there is a compelling need.

scabrero replied:
Changed to false.

Firstyear commented on server/lib/src/credential/apppwd.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+#[derive(Debug)]
+pub struct GenerateApplicationPasswordEvent {
```
Should be in idm/application.rs so that it's self-contained to show the apis and events that can occur.

scabrero replied:
Done.

Firstyear commented on server/lib/src/idm/account.rs:
```diff
@@ -697,6 +706,48 @@ impl Account {
+                    return Ok(Some(LdapBoundToken {
+                        spn: self.spn.clone(),
+                        session_id,
+                        effective_session: LdapSession::UnixBind(self.uuid),
```
Should we consider this being a different type rather than UnixBind since this session Enum is meant to encode "what kind of session this is" and this is an application password session, not a unix session.

scabrero replied:
Makes sense, I added `LdapSession::ApplicationPasswordBind`.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,715 @@
+
+#[derive(Clone)]
+struct LdapApplicationsInner {
+    set: HashMap<String, Application>,
```
What's the logic of using a string to index into this Map instead of UUID?

scabrero replied:
Mainly because `LdapApplicationAuthEvent` carries the application name extracted from the bind DN, so we can do https://github.com/scabrero/kanidm/blob/app-passwords/server/lib/src/idm/application.rs#L178

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,715 @@
+
+        let mut is_memberof = false;
+        if let Some(mut riter) = usr_entry.get_ava_as_refuuid(Attribute::MemberOf) {
+            if riter.any(|g| g == linked_group_uuid) {
```
There are very few times where you should be using `any(|| )`. Generally it's a sign that you have a the incorrect type. You should be using https://github.com/kanidm/kanidm/blob/master/server/lib/src/entry.rs#L2556 because then you can do:

```
let is_memberof = usr_entry.get_ava_refer(Attribute::MemberOf)
    .map(|member_of_set| member_of_set.contains(linked_group_uuid))
    .unwrap_or_default();
```

scabrero replied:
Fixed, thanks for the tip.

Firstyear reviewed (changes requested):
Haven't finished the full review but here is the first pass.

Firstyear commented on book/src/access_control/intro.md:
```diff
@@ -75,6 +75,7 @@ tasks within Kanidm as required.
 | `idm_schema_admins`          | add and modify elements of schema                                       |
 | `idm_service_account_admins` | create and modify service accounts                                      |
 | `idm_unix_admins`            | enable posix attributes on accounts and groups                          |
+| `idm_ldap_application_admins`| create and modify applications
```
create and modify ldap applications

scabrero replied:
Fixed.

Firstyear commented on server/lib/src/constants/schema.rs:
```diff
@@ -731,7 +752,34 @@ pub static ref SCHEMA_CLASS_PERSON_DL5: SchemaClass = SchemaClass {
+    ..Default::default()
+};
+
+pub static ref SCHEMA_CLASS_PERSON_DL6: SchemaClass = SchemaClass {
```
Minor, but just an artefact of the review taking a while is these need to target DL7 now, not DL6 https://github.com/kanidm/kanidm/blob/master/server/lib/src/server/migrations.rs#L654

scabrero replied:
Updated.

Firstyear commented on server/lib/src/idm/application.rs:
```diff
@@ -0,0 +1,741 @@
+            .collect();
+
+        app_set.map(|mut app_set| {
+            // Delay getting the inner mut (which may clone) until we know we are ok.
```
Minor, but we might consider an addition to cowcell such that we can do "update_mut" so that we don't need to clone before we swap: https://github.com/kanidm/concread/blob/master/src/cowcell/mod.rs#L183

Firstyear replied:
https://github.com/kanidm/concread/pull/116/files

Added this for you, will update the concread dep in kanidm shortly.

scabrero replied:
Great, rebased and changed to use .replace().

Firstyear commented on server/lib/src/idm/server.rs:
```diff
@@ -677,6 +693,55 @@ pub trait IdmServerTransaction<'a> {
+        ct: Duration,
+        source: Source,
+    ) -> Result<Identity, OperationError> {
+        let anon_entry = self
```
To avoid the former confusion, lets comment here that users via ldap bind are *always* only granted anonymous equivalent rights unless they auth with an api-token. (Due to ldap binds being weaker than our normal auth flow)

scabrero replied:
Agree, added.

Firstyear commented on server/lib/src/idm/server.rs:
```diff
@@ -677,6 +693,55 @@ pub trait IdmServerTransaction<'a> {
+                .as_ref(),
+        ) {
+            // Good to go
+            let limits = Limits::default();
```
I think this might be a mistake from when I added account policy, but I wonder if we should update the Limits here based on account policy from https://github.com/kanidm/kanidm/blob/master/server/lib/src/idm/server.rs#L889 ? Because these limits have impacts on search requests (and we currently don't implement paged search)

scabrero replied:
Yes, makes sense. I have updated search limits.

Firstyear commented on server/lib/src/server/mod.rs:
```diff
@@ -129,6 +129,7 @@ bitflags::bitflags! {
         const SYSTEM_CONFIG =  0b0001_0000;
         const SYNC_AGREEMENT = 0b0010_0000;
         const KEY_MATERIAL   = 0b0100_0000;
+        const APPLICATION    = 0b1000_0000;
```
I'm glad I left you with one more bit in the change flags 😅

Firstyear reviewed (commented):
This is looking much better, responded re the acp parts. I think next I need to just review that the tests cover all the situations we can think of, then I think we're almost there.

scabrero commented:
Superseded by  #2968. Thanks!

Firstyear commented:
@scabrero No, thank you! It's very appreciated :)
